### ISSUE TITLE
overrideDerivation doesn't propagate passthru which was added by `extendDerivation`

### ISSUE THREAD

**Issue by timbertson** [NixOS/nixpkgs#45933]
## Issue description

It's a weird use case (which I'll describe below), but basically: if I use `lib.overrideDerivation` on a derivation which has already been extended (via `lib.extendDerivation`), it drops the extended `passthru` attributes and reverts to the original derivation's `passthru`. This is still the case when the original derivation doesn't have any `passthru` attributes.

### Steps to reproduce

```
$ curl https://gist.github.com/timbertson/1d7beddd3a647ba8fcb589747fa81240 > testcase.nix

# sanity checking:
$ nix-instantiate --eval testcase.nix -A emptyExtended.attr
1

$ nix-instantiate --eval testcase.nix -A simpleExtended.attr
2

$ nix-instantiate --eval testcase.nix -A simpleOverride.attr
1

$ nix-instantiate --eval testcase.nix -A hasOtherAttrExtended.attr
3

# extend merges original and extended `passthru`
$ nix-instantiate --eval testcase.nix -A hasOtherAttrExtended.someOtherAttr
1

# but extend -> override ends up with only the original derivation's `passthru`
$ nix-instantiate --eval testcase.nix -A hasOtherAttrOverride.attr
error: attribute 'attr' in selection path 'hasOtherAttrOverride.attr' not found

# nix-instantiate --eval testcase.nix -A hasOtherAttrExtended.passthru
{ someOtherAttr = 1; }

$ nix-instantiate --eval testcase.nix -A hasOtherAttrOverride.someOtherAttr
1
```

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.17.17-100.fc27.x86_64, Fedora, 27 (Workstation Edition)`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0.2`
 - channels(tim): `"nixpkgs-18.09pre139886.183f56f2d62"`
 - nixpkgs: `/home/tim/.nix-defexpr/channels/nixpkgs`

# Why are you doing something so weird?

If I were hand crafting the derivations, I wouldn't. However in my case:

 - the original derivation is generated by the `opam2nix` utility, as a `.nix` file on disk. It includes one `passthru` attribute. This is a pure recipe for building a package, but requires all dependencies to be injected as package args.
 - this nix file is imported, called and extended by the nix function `buildOpamPackage` in `opam2nix-packages`. For example, it adds a `packages` attribute containing the transitive dependencies of this package.
 - In development later, I call `overrideDerivation` on the result of `buildOpamPackage` to override the `src` attribute only in development. This way I can build my local checkout without modifying the official derivation.

**Comment by timbertson** [comment]
Yes, it does.

**Comment by timbertson** [comment]
I came up with a fix that seems to work, but I don't know if it's a good idea: #45972

**Comment by dezgeg** [comment]
I wonder if `foo.overrideAttrs(old: { passthru = old.passthru // ...; })` would work instead of `extendDerivation`?

**Comment by timbertson** [comment]
I tried that too, it doesn't seem to set the actual derivation attrs (i.e. `drv.passthru.attr` would work, but not plain `drv.attr`)

**Comment by dezgeg** [comment]
I can't replicate that behaviour:
````
nix-repl> hello.foo
error: attribute 'foo' missing, at (string):1:1

nix-repl> hello2 = hello.overrideAttrs (old: { passthru = (old.passthru or {}) // { foo = (old.passthru.foo or []) ++ [1]; }; })   

nix-repl> hello2.foo
[ 1 ]

nix-repl> hello3 = hello2.overrideAttrs (old: { passthru = (old.passthru or {}) // { foo = (old.passthru.foo or []) ++ [2]; }; })     

nix-repl> hello3.foo 
[ 1 2 ]
````

**Comment by oxij** [comment]
This is not a bug, WONTFIX.

> hasOtherAttrExtended = lib.extendDerivation true { attr = 3; } hasOtherAttr;
> hasOtherAttrOverride = lib.overrideDerivation hasOtherAttrExtended (orig: {});

The problem is that you're only supposed to call `extendDerivation` over plain builtin `derivation`s. It's not designed to be called over its own result, in particular, over the results of `stdenv.mkDerivation` and `overrideDerivation`. (Nix lacks types, so the problem here is that we can't express that fact in the code explicitly without losing some good chunk of performance (`extendDerivation` is a very hot function).) Note that if you look at the source of `overrideDerivation` carefully, you'll notice that it unwraps the given derivation's original `drvAttrs`, and rewraps that into new builtin `derivation` before using `extendDerivation`. That's not an accident.

In your use case, if you want to override passthru, you are expected to do something like

> hasOtherAttrOverrideP = hasOtherAttr.overrideAttrs (orig: { passthru = orig.passthru // { attr = 3; }; });

The documentation can be improved, though, I agree.

**Comment by timbertson** [comment]
@dezgeg Hmm, I must have mistyped, your workaround does appear to work after all. Thanks for that, I will try using it as a workaround.

@oxij it was my understanding that `stdenv.mkDerivation` produces a plain derivation, because it returns you the result of calling the builtin `derivation`. What distinguishes a plain derivation from one what has been built using a utility function like this? Is it that `passthru` is an attribute given to `mkDerivation`, but that it isn't stored on the underlying derivation object?

**Comment by oxij** [comment]
> What distinguishes a plain derivation from one what has been built using a utility function like this? Is it that `passthru` is an attribute given to `mkDerivation`, but that it isn't stored on the underlying derivation object?

Yes. `stdenv.mkDerivation attrs == extendDerivation a passthru (derivation someOfTheAttrs) ~= (derivation someOfTheAttrs) // passthru` (where `passthru` is `attrs.passthru` and some other stuff like e.g. `overrideAttrs`).

**Comment by timbertson** [comment]
Is there any harm in also adding `passthru` as a native derivation attribute? It seems like it would make this work, and wouldn't affect the derivation's hash, just like any other passthru attribute. Do we _want_ this to be unsupported, or do we just happen to not support it at the moment?

**Comment by oxij** [comment]
> It seems like it would make this work, and wouldn't affect the derivation's hash

Everything you supply to `derivation` function does affect the hashes. Maybe I misunderstood what you were trying to say.

Think of `derivation attr` as being equivalent to `attr // { drvPath = drvhash attr; outPath = outhash attr; }` with side-effects.

**Comment by timbertson** [comment]
Maybe I just don't understand how `passthru` works - the general passthru mechanism adds attributes to the result without changing the derivation's output hash, doesn't it? So can't that be extended to add a `passthru` attribute as well?

**Comment by oxij** [comment]
> Maybe I just don't understand how `passthru` works - the general passthru mechanism adds attributes to the result without changing the derivation's output hash, doesn't it?

Yes.

> So can't that be extended to add a `passthru` attribute as well?

Yes. It does that too already. See the bottom of `pkgs/stdenv/generic/make-derivation.nix`.

> Is there any harm in also adding `passthru` as a native derivation attribute? 

However, now I think I see what you meant there. You see, `drv = derivation x` implies `drv.drvAttrs == x`, there's no middle ground there, either it's in the `drvAttrs` and influences the hash, or it does not and hence `overrideDerivation` doesn't know about it. Introducing a middle ground like your PR does will multiply nix memory consumption for the whole of nixpkgs.

However, I do see your point ("it's strange that `lib.overrideDerivation` is in a public API, but it does something strange to `passthru`s if you naively consider them to be part of derivations") and I do agree that something needs to be done about that. I'm inclined to say "lets just replace all `overrideDerivation` calls with `overrideAttrs`" and deprecate and then remove the former (which is best, semantically speaking), but I'm not sure if `overrideAttrs` can be made as efficient as `overrideDerivation`. I thought through four immediate ideas on how to do that, but all of them turned out to be meh. Now I have to wait for the background brain to come up with something better.
